File: examples/go/complex/cross_module.go
package main

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"net/http"
	"os/exec"
)

// ---------------------------------------------------------------------------
// cross_module.go -- Cross-file chains calling functions from all 5 existing
// Go example files: handlers.go, services.go, middleware.go, provisioner.go,
// and repository.go.
//
// All chains start from HTTP sources and flow through multiple files before
// reaching a sink. Same package (main), so direct function calls are used.
// ---------------------------------------------------------------------------

// CrossModuleHandler ties together services from all existing files.
type CrossModuleHandler struct {
	repo        *ResourceRepo
	service     *ResourceService
	provisioner *Provisioner
	configMgr   *ConfigManager
	healthCheck *HealthChecker
}

func NewCrossModuleHandler(db *sql.DB) *CrossModuleHandler {
	repo := NewResourceRepo(db)
	svc := NewResourceService(repo)
	prov := NewProvisioner()
	cm := NewConfigManager()
	return &CrossModuleHandler{
		repo:        repo,
		service:     svc,
		provisioner: prov,
		configMgr:   cm,
		healthCheck: nil,
	}
}

// ---------------------------------------------------------------------------
// CHAIN A -- Deep interprocedural: 10+ hops across 4 files
//
// Full chain (12 hops):
//   cross_module.go: ChainA_DeepFlow
//     1. r.FormValue("filter")                          -- source
//     2. parseSearchParams(filter, sortBy, limit)        -- handlers.go
//   handlers.go: parseSearchParams
//     3. returns &SearchParams{Filter: filter, ...}
//   cross_module.go: ChainA_DeepFlow (continued)
//     4. h.service.Search(params)
//   services.go: Search
//     5. buildQuery(params)
//   services.go: buildQuery
//     6. "name LIKE '%" + params.Filter + "%'"           -- string concat
//     7. returns tainted query string
//   services.go: Search (continued)
//     8. NewQueryFilter(query)
//     9. filter.Apply(params)
//     10. h.repo.FindByFilter(filter)
//   repository.go: FindByFilter
//     11. filter.Build()                                  -- services.go
//     12. r.db.Query(sqlQuery)                            -- SINK: SQL injection
//
// This reuses the same deep chain from the original files but enters through
// a new handler that adds an extra decoration step.
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainA_DeepFlow(w http.ResponseWriter, r *http.Request) {
	filter := r.FormValue("filter")
	sortBy := r.FormValue("sort")
	limit := r.FormValue("limit")

	// Hop 1: call parseSearchParams from handlers.go
	params := parseSearchParams(filter, sortBy, limit)

	// Hop 2: decorate with config (stays tainted)
	prefix := h.configMgr.Get("search.prefix")
	if prefix != nil {
		params.Filter = fmt.Sprintf("%v", prefix) + params.Filter
	}

	// Hop 3-12: service.Search -> buildQuery -> NewQueryFilter -> repo.FindByFilter -> db.Query
	results, err := h.service.Search(params)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(results)
}

// ---------------------------------------------------------------------------
// CHAIN B -- Package-level variable
//
// Full chain:
//   1. r.URL.Query().Get("config") -- source in ChainB_PackageVar
//   2. stored into package-level globalConfigData
//   3. ChainB_ConsumeConfig reads globalConfigData
//   4. h.repo.CustomQuery(clause)   -- repository.go
//   5. "SELECT * FROM resources WHERE " + whereClause -- SINK: SQL injection
//
// The package-level variable bridges two separate handler functions.
// ---------------------------------------------------------------------------

var globalConfigData string

func (h *CrossModuleHandler) ChainB_PackageVar(w http.ResponseWriter, r *http.Request) {
	// Source: user-controlled config data stored in package variable
	globalConfigData = r.URL.Query().Get("config")
	fmt.Fprintf(w, "config stored")
}

func (h *CrossModuleHandler) ChainB_ConsumeConfig(w http.ResponseWriter, r *http.Request) {
	// Read from package variable -- still tainted
	clause := "category = '" + globalConfigData + "'"
	// Sink via repository.go: CustomQuery -> db.Query
	results, err := h.repo.CustomQuery(clause)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(results)
}

// ---------------------------------------------------------------------------
// CHAIN C -- Callback / function parameter
//
// Full chain:
//   1. r.FormValue("region")                            -- source
//   2. applyTransform(region, transformRegion)           -- passes callback
//   3. transformRegion(input) returns tainted string
//   4. h.provisioner.Provision(type, transformed, name)  -- provisioner.go
//   5. NewProvisionConfig(type, region, name)             -- provisioner.go
//   6. config.BuildCommand()                              -- provisioner.go
//   7. "provision-tool --type " + type + " --region " + region + ...
//   8. p.executeProvision(cmdStr)                         -- provisioner.go
//   9. exec.Command("sh", "-c", cmdStr)                   -- SINK: cmd injection
//
// Taint flows through a function passed as a parameter (callback pattern).
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainC_Callback(w http.ResponseWriter, r *http.Request) {
	region := r.FormValue("region")
	resType := r.FormValue("type")
	name := r.FormValue("name")

	// Pass a transform function as callback -- taint must flow through
	transformed := applyTransform(region, transformRegion)

	// Sink via provisioner.go: Provision -> BuildCommand -> executeProvision -> exec.Command
	result, err := h.provisioner.Provision(resType, transformed, name)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(result)
}

func applyTransform(input string, fn func(string) string) string {
	return fn(input)
}

func transformRegion(input string) string {
	// No sanitization -- returns tainted input with a prefix
	return "region-" + input
}

// ---------------------------------------------------------------------------
// CHAIN D -- Re-exported wrapper
//
// Full chain:
//   1. r.URL.Query().Get("endpoint")                      -- source
//   2. wrapHealthCheck(endpoint)                           -- local wrapper
//   3. hc.CheckEndpoint(url)                               -- provisioner.go
//   4. http.Get(url)                                       -- SINK: SSRF
//
// AND:
//   5. r.FormValue("user")                                -- source
//   6. wrapAuditLog(user, action)                          -- local wrapper
//   7. fmt.Printf("AUDIT: user=%s action=%s\n", user, a)  -- similar to handlers.go
//   8. exec.Command("logger", "-t", "audit", msg)          -- SINK: cmd injection
//
// The wrapper functions re-export functionality from other files, adding a
// thin layer that taint must propagate through.
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainD_ReexportedWrapper(w http.ResponseWriter, r *http.Request) {
	endpoint := r.URL.Query().Get("endpoint")
	user := r.FormValue("user")
	action := r.FormValue("action")

	// Wrapper around HealthChecker.CheckEndpoint from provisioner.go
	if h.healthCheck != nil {
		status := wrapHealthCheck(h.healthCheck, endpoint)
		fmt.Fprintf(w, "health: %s\n", status)
	}

	// Wrapper around audit logging -- re-exports with exec
	wrapAuditLog(user, action)
	fmt.Fprintf(w, "audited %s %s", user, action)
}

func wrapHealthCheck(hc *HealthChecker, endpoint string) string {
	// Thin wrapper that passes taint through to provisioner.go CheckEndpoint
	status, err := hc.CheckEndpoint(endpoint)
	if err != nil {
		return "error"
	}
	return status
}

func wrapAuditLog(user, action string) {
	msg := fmt.Sprintf("user=%s action=%s", user, action)
	// SINK: command injection via logger command
	cmd := exec.Command("logger", "-t", "audit", msg)
	cmd.Run()
}

// ---------------------------------------------------------------------------
// CHAIN E -- Recursive chain
//
// Full chain:
//   1. r.URL.Query().Get("expr")                          -- source
//   2. expandExpression(expr, depth=0)                     -- recursive
//   3. expandExpression calls itself with modified expr
//   4. base case returns accumulated tainted string
//   5. h.repo.CustomQuery(expanded)                        -- repository.go
//   6. "SELECT * FROM resources WHERE " + whereClause      -- SINK: SQL injection
//
// Taint must survive recursive function calls. Each level of recursion
// appends to the tainted string without sanitizing.
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainE_Recursive(w http.ResponseWriter, r *http.Request) {
	expr := r.URL.Query().Get("expr")
	expanded := expandExpression(expr, 0)
	// Sink via repository.go: CustomQuery -> db.Query
	results, err := h.repo.CustomQuery(expanded)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(results)
}

func expandExpression(expr string, depth int) string {
	if depth >= 3 {
		return expr
	}
	// Each recursion level wraps the expression -- taint survives
	wrapped := "(" + expr + ")"
	return expandExpression(wrapped, depth+1)
}

// ---------------------------------------------------------------------------
// CHAIN F -- Cross-file struct field via middleware token
//
// Full chain:
//   1. r.Header.Get("X-Custom-Query")                     -- source
//   2. generateToken(query)                                -- middleware.go
//   3. validateToken(token)                                -- middleware.go
//   4. extractPayload(token) retrieves the tainted part
//   5. h.repo.CustomQuery(payload)                         -- repository.go
//   6. "SELECT * FROM resources WHERE " + whereClause      -- SINK: SQL injection
//
// Taint flows through the token generation from middleware.go. The payload
// portion of the token preserves the original tainted input.
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainF_MiddlewareToken(w http.ResponseWriter, r *http.Request) {
	query := r.Header.Get("X-Custom-Query")

	// Use generateToken from middleware.go -- payload is the tainted query
	token := generateToken(query)

	// Extract the payload back (it is the part before the dot)
	payload := extractPayload(token)

	// Sink via repository.go
	results, err := h.repo.CustomQuery(payload)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(results)
}

func extractPayload(token string) string {
	// Token format from middleware.go: payload.signature
	parts := splitOnDot(token)
	if len(parts) >= 1 {
		return parts[0]
	}
	return ""
}

func splitOnDot(s string) []string {
	var result []string
	current := ""
	for _, c := range s {
		if c == '.' {
			result = append(result, current)
			current = ""
		} else {
			current += string(c)
		}
	}
	if current != "" {
		result = append(result, current)
	}
	return result
}

// ---------------------------------------------------------------------------
// CHAIN G -- Service + provisioner combined chain
//
// Full chain:
//   1. json.NewDecoder(r.Body).Decode(&req)               -- source
//   2. h.service.Create(name, type, region)                -- services.go
//   3. h.repo.Create(resource)                             -- repository.go (safe)
//   4. req.Command passed to h.provisioner.Provision       -- provisioner.go
//   5. NewProvisionConfig -> BuildCommand -> executeProvision
//   6. exec.Command("sh", "-c", cmdStr)                    -- SINK: cmd injection
//
// This chain combines the service layer (for the resource) with the
// provisioner (for the command), showing taint flowing through both paths.
// ---------------------------------------------------------------------------

func (h *CrossModuleHandler) ChainG_CombinedServiceProvisioner(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Name   string `json:"name"`
		Type   string `json:"type"`
		Region string `json:"region"`
	}
	json.NewDecoder(r.Body).Decode(&req)

	// Path 1: create resource through services.go -> repository.go (parameterized, safe)
	_, err := h.service.Create(req.Name, req.Type, req.Region)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}

	// Path 2: provision through provisioner.go (tainted command injection)
	result, err := h.provisioner.Provision(req.Type, req.Region, req.Name)
	if err != nil {
		http.Error(w, err.Error(), 500)
		return
	}
	json.NewEncoder(w).Encode(result)
}

File: examples/python/complex/cross_module.py
"""
Cross-module flow test file.

Imports from ALL 5 existing Python example files and creates 5 distinct
cross-file vulnerability chains. Each chain uses Flask route handlers
as entry points (with request.args/form/json as sources) and calls
imported functions that reach sinks in other files.
"""

import os
import sys
import sqlite3
import subprocess
from flask import Flask, request, jsonify

from database import (
    DatabaseConnection,
    QueryBuilder,
    UserRepository,
    DocumentRepository,
    AuditLog,
)
from services import (
    SearchQuery,
    UserService,
    DocumentService,
    AdminService,
    NotificationService,
)
from utils import (
    sanitize_input,
    ConfigManager,
)
from ml_pipeline import (
    ModelConfig,
    ExternalModelRunner,
    PipelineManager,
)

app = Flask(__name__)
db = DatabaseConnection("app.db")


# ============================================================
# Chain A: Deep Interprocedural SQL Injection (10+ hops, 4 files)
#
# Full chain (cross_module.py -> services.py -> database.py):
#   1.  search_handler()             [cross_module.py] - Flask route, reads request.args
#   2.  enrich_search_params()       [cross_module.py] - adds metadata to query
#   3.  dispatch_search()            [cross_module.py] - selects search strategy
#   4.  execute_user_search()        [cross_module.py] - creates SearchQuery object
#   5.  SearchQuery.__init__()       [services.py]     - stores raw_query
#   6.  SearchQuery.build_filter()   [services.py]     - builds SQL WHERE clause
#   7.  UserRepository.__init__()    [database.py]     - stores db ref
#   8.  UserRepository.find_by_filter() [database.py]  - calls get_sql()
#   9.  SearchQuery.get_sql()        [services.py]     - returns unsanitized SQL
#   10. cursor.execute()             [database.py]     - executes tainted SQL
#
# Tests: deep interprocedural tracking across 3 files with 10+ hops,
#   object construction, method chains, and field propagation.
# ============================================================

# VULN: Chain A - SQL Injection via deep cross-file call chain
# Source: request.args["query"] in search_handler
# Sink: cursor.execute in database.py


def enrich_search_params(raw_query, context):
    """Adds search context metadata. Passes query through unchanged."""
    enriched = raw_query
    if context.get("fuzzy"):
        enriched = enriched + "%"
    return enriched


def dispatch_search(query_text, sort_field, limit, strategy="default"):
    """Selects search strategy and delegates. Query flows through."""
    if strategy == "advanced":
        return execute_user_search(query_text, sort_field, limit)
    return execute_user_search(query_text, sort_field, limit)


def execute_user_search(query_text, sort_field, limit):
    """Creates SearchQuery and delegates to UserRepository.
    query_text flows into SearchQuery -> build_filter -> get_sql -> cursor.execute.
    """
    search_query = SearchQuery(query_text, sort_field, limit)
    search_query.validate()
    search_query.build_filter()
    repo = UserRepository(db)
    return repo.find_by_filter(search_query)


@app.route("/api/cross/search")
def search_handler():
    """Chain A entry point. User input flows through 4 functions in this file,
    then into services.py and database.py for SQL execution.
    """
    raw_query = request.args["query"]
    sort_by = request.args.get("sort", "name")
    limit = request.args.get("limit", "100")
    context = {"fuzzy": True, "source": "api"}
    enriched = enrich_search_params(raw_query, context)
    results = dispatch_search(enriched, sort_by, limit, strategy="advanced")
    return jsonify(results)


# ============================================================
# Chain B: Module-Level Variable Command Injection (2 files)
#
# Full chain (cross_module.py -> ml_pipeline.py):
#   1. capture_handler()             [cross_module.py] - Flask route, reads request.form
#   2. _pending_script_path          [cross_module.py] - module-level variable (bridge)
#   3. execute_pending_handler()     [cross_module.py] - reads module var
#   4. PipelineManager.__init__()    [ml_pipeline.py]  - creates manager
#   5. PipelineManager.run_shell_pipeline() [ml_pipeline.py] - subprocess.run(shell=True)
#
# Tests: module-level variable bridging (bridgeable_vars), cross-function
#   flow via shared module state, and cross-file sink invocation.
# ============================================================

# VULN: Chain B - Command Injection via module-level variable
# Source: request.form["script"] in capture_handler
# Sink: subprocess.run(shell=True) in ml_pipeline.py

_pending_script_path = None


@app.route("/api/cross/capture", methods=["POST"])
def capture_handler():
    """Stores user-controlled script path in module-level variable."""
    global _pending_script_path
    _pending_script_path = request.form["script"]
    return jsonify({"status": "captured"})


@app.route("/api/cross/execute-pending", methods=["POST"])
def execute_pending_handler():
    """Reads the module-level variable and passes to PipelineManager."""
    global _pending_script_path
    if _pending_script_path is None:
        return jsonify({"error": "nothing pending"}), 400
    manager = PipelineManager()
    result = manager.run_shell_pipeline(_pending_script_path)
    _pending_script_path = None
    return jsonify({"output": result})


# ============================================================
# Chain C: Callback/Event Command Injection (3 files)
#
# Full chain (cross_module.py -> services.py):
#   1. trigger_handler()             [cross_module.py] - Flask route, reads request.json
#   2. _on_maintenance_event()       [cross_module.py] - callback function
#   3. AdminService.__init__()       [services.py]     - stores db ref
#   4. AdminService.run_maintenance() [services.py]    - subprocess.run(shell=True)
#
# Tests: callback function stored in dict, invoked with tainted input,
#   delegates to imported class method that reaches sink.
# ============================================================

# VULN: Chain C - Command Injection via callback mechanism
# Source: request.json["payload"] in trigger_handler
# Sink: subprocess.run(shell=True) in services.py AdminService.run_maintenance

_event_handlers = {}


def _on_maintenance_event(task_name):
    """Callback that delegates to AdminService.run_maintenance."""
    svc = AdminService(db)
    return svc.run_maintenance(task_name)


def _on_notification_event(address):
    """Callback that delegates to NotificationService.send_email."""
    notifier = NotificationService()
    notifier.send_email(address, "Event triggered", "Automated notification")
    return {"sent": True}


def register_event_handlers():
    """Registers callbacks into a dictionary for later dispatch."""
    _event_handlers["maintenance"] = _on_maintenance_event
    _event_handlers["notification"] = _on_notification_event


register_event_handlers()


@app.route("/api/cross/trigger", methods=["POST"])
def trigger_handler():
    """Reads user input and dispatches to registered callback."""
    event_type = request.json["type"]
    payload = request.json["payload"]
    handler = _event_handlers.get(event_type)
    if handler is None:
        return jsonify({"error": "unknown event"}), 400
    result = handler(payload)
    return jsonify(result)


# ============================================================
# Chain D: Re-exported Wrapper (3 files)
#
# Full chain (cross_module.py -> services.py -> os.system):
#   1. notify_handler()              [cross_module.py] - Flask route, reads request.form
#   2. send_single_notification()    [cross_module.py] - wrapper function
#   3. NotificationService.__init__() [services.py]    - creates service
#   4. NotificationService.send_email() [services.py]  - os.system(cmd)
#
# Also sub-chain D2 through ml_pipeline.py:
#   1. model_handler()               [cross_module.py] - Flask route, reads request.form
#   2. prepare_and_run_model()       [cross_module.py] - wrapper function
#   3. ExternalModelRunner.__init__() [ml_pipeline.py] - stores config
#   4. ExternalModelRunner.execute()  [ml_pipeline.py] - os.system(cmd)
#
# Tests: import resolution for re-exported/wrapped functions.
# ============================================================

# VULN: Chain D - Command Injection via re-exported wrapper
# Source: request.form["recipient"] in notify_handler
# Sink: os.system in services.py NotificationService.send_email


def send_single_notification(recipient, subject):
    """Wrapper around NotificationService.send_email."""
    notifier = NotificationService()
    notifier.send_email(recipient, subject, "Automated message body")
    return {"status": "sent", "to": recipient}


@app.route("/api/cross/notify", methods=["POST"])
def notify_handler():
    """Chain D1 entry point. User input flows through wrapper to sink."""
    recipient = request.form["recipient"]
    subject = request.form.get("subject", "Alert")
    result = send_single_notification(recipient, subject)
    return jsonify(result)


def prepare_and_run_model(model_path, model_name):
    """Wrapper that creates ExternalModelRunner and calls execute."""
    config = ModelConfig(model_path, model_name)
    config.validate()
    runner = ExternalModelRunner(config)
    resolved = runner.prepare()
    return runner.execute(resolved)


@app.route("/api/cross/run-model", methods=["POST"])
def model_handler():
    """Chain D2 entry point. User input flows to ml_pipeline.py sink."""
    model_path = request.form["model_path"]
    model_name = request.form.get("model_name", "default")
    result = prepare_and_run_model(model_path, model_name)
    return jsonify(result)


# ============================================================
# Chain E: Recursive SQL Injection (3 files)
#
# Full chain (cross_module.py -> database.py):
#   1. audit_handler()               [cross_module.py] - Flask route, reads request.args
#   2. expand_query_recursive()      [cross_module.py] - recursive function
#   3. execute_raw_query()           [cross_module.py] - calls AuditLog.search_logs
#   4. AuditLog.__init__()          [database.py]     - stores db ref
#   5. AuditLog.search_logs()       [database.py]     - cursor.execute(string concat)
#
# Tests: recursive call handling where taint survives through
#   recursive invocations and eventually reaches a cross-file sink.
# ============================================================

# VULN: Chain E - SQL Injection via recursive query expansion
# Source: request.args["filter"] in audit_handler
# Sink: cursor.execute in database.py AuditLog.search_logs


def expand_query_recursive(query_fragment, depth=0):
    """Recursively expands query fragments. Taint survives recursion."""
    if depth > 5:
        return query_fragment
    marker = "{{sub:"
    idx = query_fragment.find(marker)
    if idx == -1:
        return query_fragment
    end_idx = query_fragment.find("}}", idx + len(marker))
    if end_idx == -1:
        return query_fragment
    sub_value = query_fragment[idx + len(marker):end_idx]
    expanded_sub = expand_query_recursive(sub_value, depth + 1)
    result = query_fragment[:idx] + expanded_sub + query_fragment[end_idx + 2:]
    return result


def execute_raw_query(filter_text):
    """Passes the expanded query to AuditLog.search_logs."""
    audit = AuditLog(db)
    return audit.search_logs(filter_text)


@app.route("/api/cross/audit")
def audit_handler():
    """Chain E entry point. Recursively expands user input, then executes."""
    raw_filter = request.args["filter"]
    expanded = expand_query_recursive(raw_filter)
    results = execute_raw_query(expanded)
    return jsonify(results)


# ============================================================
# Safe chain: sanitized input does NOT reach sink
# ============================================================

@app.route("/api/cross/safe")
def safe_handler():
    """Safe variant: input goes through sanitize_input from utils.py."""
    raw = request.args["q"]
    cleaned = sanitize_input(raw)
    conn = db.get_connection()
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM items WHERE name = ?", (cleaned,))
    return jsonify([dict(r) for r in cursor.fetchall()])


if __name__ == "__main__":
    app.run(debug=True, host="0.0.0.0", port=5002)
